File: miniflow-engine/hotkey.py
"""
Hotkey config — TWO triggers:

  "dictation"  → grammar-correct the transcript and type it into the focused app
  "command"    → run the transcript through the full agent loop (tools, connectors)

Stored at ~/miniflow/hotkey.json. Read by:
 - The native helper binary (miniflow-fn-helper) on startup and on SIGHUP
 - The Electron main process

New schema (v2):
    {
      "dictation": { "mode": "hold_to_talk", "modifier": "fn",     "key": null },
      "command":   { "mode": "hold_to_talk", "modifier": "option", "key": "space" }
    }

Legacy schema (v1) is auto-migrated: if the root has "mode" / "modifier" / "key"
those fields are moved under "dictation" and a sensible "command" default is added.
"""

from __future__ import annotations

import copy
import json
import logging
import stat
from pathlib import Path
from typing import Any

log = logging.getLogger("hotkey")

HOTKEY_FILE = Path.home() / "miniflow" / "hotkey.json"

VALID_MODES = {"hold_to_talk", "press_to_toggle"}
VALID_MODIFIERS = {"fn", "cmd", "option", "control", "shift", "globe", None}
VALID_KEYS = {
    None,
    "space", "return", "tab", "escape", "delete",
    "up", "down", "left", "right",
    *(chr(c) for c in range(ord("a"), ord("z") + 1)),
    *(str(n) for n in range(10)),
    *(f"f{n}" for n in range(1, 13)),
}

DEFAULT_DICTATION = {"mode": "hold_to_talk",    "modifier": "fn",     "key": None}
# Command mode defaults to press-to-toggle: tap once to start, tap again to stop
# speaking. This lets the user's hand return to the keyboard while the LLM thinks.
DEFAULT_COMMAND   = {"mode": "press_to_toggle", "modifier": "option", "key": "space"}

DEFAULT_HOTKEY: dict[str, Any] = {
    "dictation": dict(DEFAULT_DICTATION),
    "command":   dict(DEFAULT_COMMAND),
}


# ── File IO ───────────────────────────────────────────────────────────────────

def _write(data: dict):
    HOTKEY_FILE.parent.mkdir(exist_ok=True)
    HOTKEY_FILE.write_text(json.dumps(data, indent=2))
    HOTKEY_FILE.chmod(stat.S_IRUSR | stat.S_IWUSR)


def _read_raw() -> dict:
    try:
        if HOTKEY_FILE.exists():
            return json.loads(HOTKEY_FILE.read_text())
    except Exception as e:
        log.warning(f"hotkey.json unreadable, using defaults: {e}")
    return {}


def _migrate(raw: dict) -> dict:
    """Accept old-flat-schema or new-nested-schema and return new-nested."""
    if "dictation" in raw or "command" in raw:
        out = copy.deepcopy(DEFAULT_HOTKEY)
        if isinstance(raw.get("dictation"), dict):
            out["dictation"] = {**DEFAULT_DICTATION, **raw["dictation"]}
        if "command" in raw:
            if raw["command"] is None:
                out["command"] = None
            elif isinstance(raw["command"], dict):
                out["command"] = {**DEFAULT_COMMAND, **raw["command"]}
        return out
    # Legacy flat schema — migrate top-level fields into "dictation"
    legacy = {k: raw[k] for k in ("mode", "modifier", "key") if k in raw}
    return {
        "dictation": {**DEFAULT_DICTATION, **legacy},
        "command":   dict(DEFAULT_COMMAND),
    }


# ── Validation ────────────────────────────────────────────────────────────────

def _validate_single(hk: dict, label: str) -> dict:
    mode = hk.get("mode", "hold_to_talk")
    modifier = hk.get("modifier")
    key = hk.get("key")
    if mode not in VALID_MODES:
        raise ValueError(f"[{label}] invalid mode: {mode!r}")
    if modifier not in VALID_MODIFIERS:
        raise ValueError(f"[{label}] invalid modifier: {modifier!r}")
    if key not in VALID_KEYS:
        raise ValueError(f"[{label}] invalid key: {key!r}")
    if modifier is None and key is None:
        raise ValueError(f"[{label}] at least one of modifier/key must be set")
    if key is None and mode == "press_to_toggle":
        raise ValueError(f"[{label}] press_to_toggle requires a non-modifier key")
    return {"mode": mode, "modifier": modifier, "key": key}


def validate(hk: dict) -> dict:
    """Validate the full two-hotkey config, returning a normalized copy."""
    out: dict = {}
    if "dictation" in hk:
        out["dictation"] = _validate_single(hk["dictation"], "dictation")
    else:
        out["dictation"] = dict(DEFAULT_DICTATION)
    if "command" in hk:
        if hk["command"] is None:
            out["command"] = None
        else:
            out["command"] = _validate_single(hk["command"], "command")
    else:
        out["command"] = dict(DEFAULT_COMMAND)
    # Don't allow the two hotkeys to collide
    d, c = out["dictation"], out["command"]
    if c and d["modifier"] == c["modifier"] and d["key"] == c["key"]:
        raise ValueError("Dictation and command hotkeys must differ")
    return out


# ── Public API ────────────────────────────────────────────────────────────────

def get_hotkey() -> dict:
    return _migrate(_read_raw()) or copy.deepcopy(DEFAULT_HOTKEY)


def set_hotkey(hk: dict) -> dict:
    normalized = validate(_migrate(hk))
    _write(normalized)
    log.info(f"Hotkey updated: {normalized}")
    _signal_helper()
    return normalized


def reset_hotkey() -> dict:
    default = copy.deepcopy(DEFAULT_HOTKEY)
    _write(default)
    log.info("Hotkey reset to defaults")
    _signal_helper()
    return default


def _signal_helper():
    """SIGHUP the native helper so it reloads hotkey.json without a restart."""
    import os, signal
    pidfile = Path.home() / "miniflow" / "miniflow-fn-helper.pid"
    if not pidfile.exists():
        return
    try:
        pid = int(pidfile.read_text().strip())
        os.kill(pid, signal.SIGHUP)
        log.info(f"Sent SIGHUP to native helper pid={pid}")
    except Exception as e:
        log.warning(f"Could not SIGHUP helper: {e}")

File: native-helper/src/hotkey.rs
//! Load `~/miniflow/hotkey.json`. Now supports TWO triggers:
//!   - dictation: grammar-corrected typing path
//!   - command:   full agent / tool-calling path
//!
//! Schema:
//!   {
//!     "dictation": { "mode": "hold_to_talk", "modifier": "fn",     "key": null },
//!     "command":   { "mode": "hold_to_talk", "modifier": "option", "key": "space" }
//!   }
//!
//! Legacy flat schema ({ mode, modifier, key } at root) is auto-migrated to
//! dictation + a default command binding.

use std::fs;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Mode { HoldToTalk, PressToToggle }

impl Default for Mode { fn default() -> Self { Mode::HoldToTalk } }

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Modifier { Fn, Cmd, Option, Control, Shift, Globe }

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct HotkeyBinding {
    #[serde(default)]
    pub mode: Mode,
    #[serde(default = "default_modifier")]
    pub modifier: Option<Modifier>,
    #[serde(default)]
    pub key: Option<String>,
}

fn default_modifier() -> Option<Modifier> { Some(Modifier::Fn) }

impl Default for HotkeyBinding {
    fn default() -> Self {
        Self { mode: Mode::HoldToTalk, modifier: Some(Modifier::Fn), key: None }
    }
}

impl HotkeyBinding {
    pub fn is_modifier_only(&self) -> bool {
        self.key.is_none() && self.modifier.is_some()
    }
}

#[derive(Debug, Clone)]
pub struct HotkeyConfig {
    pub dictation: HotkeyBinding,
    pub command: Option<HotkeyBinding>,
}

impl Default for HotkeyConfig {
    fn default() -> Self {
        Self {
            dictation: HotkeyBinding::default(),
            // Command mode: press-to-toggle (tap once to start, tap again to stop).
            // Matches miniflow-engine/hotkey.py DEFAULT_COMMAND.
            command: Some(HotkeyBinding {
                mode: Mode::PressToToggle,
                modifier: Some(Modifier::Option),
                key: Some("space".into()),
            }),
        }
    }
}

#[derive(Debug, Deserialize)]
struct RawNested {
    dictation: Option<HotkeyBinding>,
    command: Option<HotkeyBinding>,
}

pub fn config_path() -> PathBuf {
    let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("."));
    home.join("miniflow").join("hotkey.json")
}

pub fn load() -> HotkeyConfig {
    let raw = match fs::read_to_string(config_path()) {
        Ok(s) => s,
        Err(_) => return HotkeyConfig::default(),
    };
    // Try the new nested format first
    if let Ok(parsed) = serde_json::from_str::<RawNested>(&raw) {
        if parsed.dictation.is_some() || parsed.command.is_some() {
            let def = HotkeyConfig::default();
            return HotkeyConfig {
                dictation: parsed.dictation.unwrap_or(def.dictation),
                command: parsed.command.or(def.command),
            };
        }
    }
    // Fall back to the legacy flat schema ({ mode, modifier, key })
    if let Ok(flat) = serde_json::from_str::<HotkeyBinding>(&raw) {
        let def = HotkeyConfig::default();
        return HotkeyConfig { dictation: flat, command: def.command };
    }
    eprintln!("[helper] hotkey.json parse failed; using defaults");
    HotkeyConfig::default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_matches_python_defaults() {
        let c = HotkeyConfig::default();
        assert_eq!(c.dictation.modifier, Some(Modifier::Fn));
        assert_eq!(c.dictation.key, None);
        let cmd = c.command.expect("command default present");
        assert_eq!(cmd.modifier, Some(Modifier::Option));
        assert_eq!(cmd.key.as_deref(), Some("space"));
        assert_eq!(cmd.mode, Mode::PressToToggle);
    }

    #[test]
    fn parse_nested() {
        let js = r#"{
          "dictation": {"mode":"hold_to_talk","modifier":"fn","key":null},
          "command":   {"mode":"hold_to_talk","modifier":"cmd","key":"d"}
        }"#;
        let c = serde_json::from_str::<RawNested>(js).unwrap();
        assert_eq!(c.dictation.unwrap().modifier, Some(Modifier::Fn));
        assert_eq!(c.command.unwrap().key.as_deref(), Some("d"));
    }

    #[test]
    fn parse_flat_legacy() {
        let js = r#"{"mode":"hold_to_talk","modifier":"control","key":null}"#;
        let b: HotkeyBinding = serde_json::from_str(js).unwrap();
        assert!(b.is_modifier_only());
        assert_eq!(b.modifier, Some(Modifier::Control));
    }
}
